### PR TITLE
mftrace: init at 1.2.20

### LINKED ISSUE #117876: mftrace
**Project description**
From the project website: "mftrace is a small Python program that lets you trace a TeX bitmap font into a PFA or PFB font (A PostScript Type1 Scalable Font) or TTF (TrueType) font."

I came across it while trying to compile the [Comprehensive LaTeX Symbol list](https://ctan.org/tex-archive/info/symbols/comprehensive). It doesn't seem to be used a lot, but it seems nice to have, and I have just packaged it for myself anyway.

Working packaging attempt [here](https://github.com/NixOS/nixpkgs/compare/master...xworld21:mftrace), just two notes:
- mftrace needs texlive at runtime only (it finds fonts via `kpsewhich` and uses metafont binaries), I have left it out so that it works with any installed texlive, but it's also broken if there is none, is that ok?
- I have not added a maintainer!

**Metadata**

* homepage URL: https://lilypond.org/mftrace/
* source URL: https://github.com/hanwen/mftrace
* licenses: gpl2-only, free
* platforms: unix, windows (judging from the source – I have not tested it)

### PR DESCRIPTION
**Pull request by xworld21**
###### Motivation for this change
See my packaging request in #117876.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [X] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: maintainers/maintainer-list.nix [hunk 1]
```diff
@@ -10604,6 +10604,12 @@
     githubId = 11824817;
     name = "Marti Serra";
   };
+  xworld21 = {
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10607]: Please put this into a separate commit with the message ``maintainers: add xworld21`` and sort it alphabetically.

File: maintainers/maintainer-list.nix [hunk 2]
```diff
@@ -10604,6 +10604,12 @@
...
+    email = "xworld21@users.sf.net";
```
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:10608]: The anonymous github mail would be better incase sourceforge changes anything and we can't reach you or a mail directly from your mail provider.

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
+{ stdenv, fetchFromGitHub, lib, makeWrapper, autoconf, python3, fontforge, potrace }:
+stdenv.mkDerivation rec {
```
Comment (SuperSandro2000) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:2]:
```suggestion
{ stdenv, fetchFromGitHub, lib, makeWrapper, autoconf, python3, fontforge, potrace }:

stdenv.mkDerivation rec {
```

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+  # at runtime, mftrace needs:
+  # - a tex distribution containing the fonts and the binaries
+  #   kpsewich, mf (metafont), t1asm, mktextfm
+  # - one of potrace (preferred), autotrace
+  # - one of fontforge (preferred), ghostscript
+  extraBuildInputs = [ fontforge potrace ];
+
+  preConfigure = ''
+    autoconf
+  '';
```
Comment (SuperSandro2000) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:33]: We should probably use autoreconfHook here to prevent issues when updating autoconf etc.

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  postInstall = lib.optionalString (extraBuildInputs != [ ]) ''
```
Comment (SuperSandro2000) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:35]: This can't be overriden right now.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix (hunk 2 above)
Comment (xworld21) [reply maintainers/maintainer-list.nix:10608]: Huh, actually, it looks like my old alias has already stopped working. Thanks for pointing this out, I'll change email.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 5 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:35]: At last, I have read the source of `mkDerivation`, I get it now. That doesn't work as I thought.

Would passing `extraBuildInputs` as an argument be reasonable?

Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:35]: Just for the record, I tried implementing a `passthru.withEnv` function that wraps a minimal build (i.e. without `postInstall`) with whatever you pass it. I took inspiration from some `passthru.withPlugins` I have seen around. It worked very well, but it also felt overkill (I had to create an `mftraceMinimal` as well to keep things properly overrideable).

Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:32]: @SuperSandro2000 is it ok if I hardcode `potrace` and `fontforge` inside `postInstall`, say after checking that they are not `null`? Then one can write e.g. `fontforge = ghostscript` or `fontforge = null` if they want to override those.

On a related note, I'll try to make make `mftrace` compatible with `texlive.combine`, so if one wants to bind `mftrace` with a particular `texlive` instance they can use that mechanism. It sounds better than what I was trying to do here.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix]: I changed the mechanism so that the dependencies can be overridden via e.g. `fontforge = ghostscript` (the options are described in a comment). I have also made it `texlive.combine`-able.

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (xworld21) [reply maintainers/maintainer-list.nix:10607]: I believe I put myself in the right spot (between `xv*` and `xwv`), it's the people after `zzamboni` who are not in order! I will fix the commit message in the final squash.

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  buildInputs = [ fontforge potrace ];
+
+  postInstall = ''
+    wrapProgram $out/bin/mftrace --prefix PATH : ${lib.makeBinPath buildInputs}
+  '';
+
+  passthru.tlType = "bin";
+  passthru.pkgs = [ self ] ++ (builtins.concatMap (builtins.getAttr "pkgs")
+    (with texlive; [ kpathsea t1utils metafont ]));
```
Comment (SuperSandro2000) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:47]: I am not sure about this as I don't see such pattern very often.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 5 above)
Comment (SuperSandro2000) [reply pkgs/tools/typesetting/tex/mftrace/default.nix]: > after checking that they are not null

You don't need to check if packages are not null in inputs and overriding packages to null in inputs is not encouraged.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 6 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:47]: It is supposed to pull all the transitive dependencies of those three packages (based on how it's done in `flatDeps` inside `texlive/default.nix`). The only other packages with `tlType`/`pkgs` are `auctex` and `noweb`, and they do not specify dependencies, so it is indeed a new thing. @veprbl would you be able to confirm that I got this right?

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 5 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix]: Indeed, I noticed that `makeBinPath` works just fine with `null`. The new code does not check for it.

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  nativeBuildInputs = [ makeWrapper autoreconfHook python3 potrace ];
+
+  buildInputs = [ fontforge potrace ];
+
+  postInstall = ''
+    wrapProgram $out/bin/mftrace --prefix PATH : ${lib.makeBinPath buildInputs}
+  '';
+
+  passthru.tlType = "bin";
```
Comment (veprbl) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:45]: mftrace doesn't provide any texmf files, so I don't see why would you want to embed it into texlive. It should be perfectly functional as a standalone program.

File: pkgs/tools/typesetting/tex/mftrace/default.nix [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+  meta = with lib; {
+    description = "Scalable PostScript Fonts for MetaFont";
+    longDescription = ''
+      mftrace is a small Python program that lets you trace a TeX bitmap
+      font into a PFA or PFB font (A PostScript Type1 Scalable Font) or
+      TTF (TrueType) font.
+    '';
+    homepage = "https://lilypond.org/mftrace/";
+    license = with licenses; [ gpl2Only free ];
```
Comment (veprbl) [inline pkgs/tools/typesetting/tex/mftrace/default.nix:57]:
```suggestion
    license = with licenses; [ gpl2Only mit ];
```

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 7 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:45]: > It should be perfectly functional as a standalone program.

That's the thing, it doesn't work at all without t1utils and metafont. I am happy to drop this change, but, a last defence: I wrote this in order to let `texlive.combine` resolve mftrace's dependencies at installation time, instead of having the user figuring it out. The stronger approach would be to specify a custom combination in the `buildInputs`, but it cannot work due to kpathsea shenanigans. This is the next best thing as far as I can tell.

My underlying request is: the packages (mftrace, LaTeXML, possibly others?) that have a dependency on texlive, but cannot use `buildInputs`/`makeWrapper`, should be made able to pass their dependencies to `texlive.combine`. This orthogonal to providing texmf files, I think.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 8 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix]: Good catch. For the record, texlive lists xdvi (the origin of the piece of code in mftrace) as "Free license not otherwise listed", but indeed the text matches MIT word for word except for "the authors". I didn't know that.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 7 above)
Comment (veprbl) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:45]: I don't see what `texlive.combine` brings over the standard PATH combination mechanizm. This works fine:
```
nix-shell -I nixpkgs=`pwd`/nixpkgs -p mftrace -p texlive.combined.scheme-small --pure --run "mftrace cmr10"
```
But, of course, if you we can keep `tlType` if you want to run it another way.

**Review by veprbl — APPROVED** [review]

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 6 above)
Comment (veprbl) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:47]: Yes, this is how texlive "packages" work. I would write `kpathsea.pkgs ++ t1utils.pkgs ++ metafont.pkgs`, but it's just my preference.

File: pkgs/tools/typesetting/tex/mftrace/default.nix (hunk 7 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/mftrace/default.nix:45]: Thanks! I suppose I am to blame – I start from collection-basic, not scheme-small, and get dependency errors quite regularly.

Given that `combine` is not your standard `buildEnv`, something I only understood a few days ago, I will add a warning that `texlive.combine { mftrace ... }` is an experimental usage (at least until #118745 is fixed, and depending on how the fix looks like).

File: maintainers/maintainer-list.nix (hunk 1 above)
Comment (SuperSandro2000) [reply maintainers/maintainer-list.nix:10641]: Yeah some entries slipped through.
